Issue with random number generation for jitter.

```
panic: invalid argument to Int63n

goroutine 26 [running]:
math/rand.(*Rand).Int63n(0xc00009c420, 0x8000000000000001, 0x3ff0000000000000)
	/usr/local/Cellar/go/1.13.6/libexec/src/math/rand/rand.go:111 +0x11d
math/rand.Int63n(...)
	/usr/local/Cellar/go/1.13.6/libexec/src/math/rand/rand.go:319
github.com/dogmatiq/linger.Rand(0x0, 0x8000000000000000, 0x3f847ae147ae147b)
	/Users/james/go/pkg/mod/github.com/dogmatiq/linger@v0.2.0/jitter.go:42 +0x55
github.com/dogmatiq/linger.FullJitter(0x8000000000000000, 0x0)
	/Users/james/go/pkg/mod/github.com/dogmatiq/linger@v0.2.0/jitter.go:14 +0x34
github.com/dogmatiq/linger/backoff.WithTransforms.func1(0x0, 0x0, 0x28, 0x4505013)
	/Users/james/go/pkg/mod/github.com/dogmatiq/linger@v0.2.0/backoff/strategy.go:71 +0x84
github.com/dogmatiq/linger/backoff.(*Counter).Fail(0xc0001f6798, 0x0, 0x0, 0x0)
	/Users/james/go/pkg/mod/github.com/dogmatiq/linger@v0.2.0/backoff/counter.go:40 +0x54
github.com/dogmatiq/linger/backoff.(*Counter).Sleep(0xc0001f6798, 0x485ec80, 0xc0001f67c0, 0x0, 0x0, 0x24, 0x0)
	/Users/james/go/pkg/mod/github.com/dogmatiq/linger@v0.2.0/backoff/counter.go:53 +0x43
github.com/dogmatiq/infix/persistence/provider/sql.(*cursor).Next(0xc0001f6780, 0x485ec80, 0xc0001f67c0, 0x0, 0x0, 0x0)
	/Users/james/grit/github.com/dogmatiq/infix/persistence/provider/sql/stream.go:214 +0x238
github.com/dogmatiq/infix/app/projection.(*Projector).consumeNext(0xc0001d0000, 0x485ec80, 0xc000134140, 0x4857480, 0xc0001f6780, 0x0, 0x0)
	/Users/james/grit/github.com/dogmatiq/infix/app/projection/projector.go:181 +0x71
github.com/dogmatiq/infix/app/projection.(*Projector).consume(0xc0001d0000, 0x485ec80, 0xc000134140, 0x0, 0x0)
	/Users/james/grit/github.com/dogmatiq/infix/app/projection/projector.go:138 +0x553
github.com/dogmatiq/infix/app/projection.(*Projector).Run(0xc0001d0000, 0x485ec80, 0xc000134140, 0xc000168088, 0xc00007a838)
	/Users/james/grit/github.com/dogmatiq/infix/app/projection/projector.go:70 +0x137
github.com/dogmatiq/infix.(*Engine).streamEvents.func1(0xc00004e768, 0xc00004e798)
	/Users/james/grit/github.com/dogmatiq/infix/app.go:57 +0x3f
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc000142270, 0xc0001324e0)
	/Users/james/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/Users/james/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66
```

--- a/backoff/strategy.go
+++ b/backoff/strategy.go
@@ -70,7 +70,7 @@ func Linear(unit time.Duration) Strategy {
 		delay := mult * unit
 
 		// Overflow check: If delay is negative, there was clearly an overflow
-		// because both n and mult are positive.
+		// because both unit and mult are positive.
 		if delay < 0 {
 			return linger.MaxDuration
 		}

--- a/backoff/strategy.go
+++ b/backoff/strategy.go
@@ -35,12 +35,18 @@ func Exponential(unit time.Duration) Strategy {
 		panic("the unit duration must be postive")
 	}
 
-	u := unit.Seconds()
+	u := float64(unit)
 
 	return func(_ error, n uint) time.Duration {
 		scale := math.Pow(2, float64(n))
-		seconds := u * scale
-		return linger.FromSeconds(seconds)
+		nanos := u * scale
+
+		// Overflow check.
+		if nanos >= float64(linger.MaxDuration) {
+			return linger.MaxDuration
+		}
+
+		return time.Duration(nanos)
 	}
 }
 
@@ -55,8 +61,27 @@ func Constant(d time.Duration) Strategy {
 //
 // The unit delay is multiplied by the number of successive failures.
 func Linear(unit time.Duration) Strategy {
+	if unit <= 0 {
+		panic("the unit duration must be postive")
+	}
+
 	return func(_ error, n uint) time.Duration {
-		return time.Duration(n+1) * unit
+		mult := time.Duration(n) + 1
+		delay := mult * unit
+
+		// Overflow check: If delay is negative, there was clearly an overflow
+		// because both n and mult are positive.
+		if delay < 0 {
+			return linger.MaxDuration
+		}
+
+		// Overflow check: Dividing the delay by the unit value should give us
+		// back the multiplier that we used. If not, there was an overflow.
+		if delay/unit != mult {
+			return linger.MaxDuration
+		}
+
+		return delay
 	}
 }
 

--- a/backoff/strategy_test.go
+++ b/backoff/strategy_test.go
@@ -28,6 +28,15 @@ var _ = Describe("func Exponential()", func() {
 			Exponential(-1)
 		}).To(Panic())
 	})
+
+	It("does not overflow the time.Duration type", func() {
+		strategy := Exponential(1)
+
+		Expect(strategy(nil, 62)).To(Equal(time.Duration(4611686018427387904)))
+		Expect(strategy(nil, 63)).To(Equal(linger.MaxDuration))
+		Expect(strategy(nil, 100)).To(Equal(linger.MaxDuration))
+	})
+
 })
 
 var _ = Describe("func Constant()", func() {
@@ -45,6 +54,28 @@ var _ = Describe("func Linear()", func() {
 
 		Expect(strategy(nil, 4)).To(Equal(15 * time.Second))
 		Expect(strategy(nil, 5)).To(Equal(18 * time.Second))
+	})
+
+	It("panics if the unit is zero", func() {
+		Expect(func() {
+			Linear(0)
+		}).To(Panic())
+	})
+
+	It("panics if the unit is negative", func() {
+		Expect(func() {
+			Linear(-1)
+		}).To(Panic())
+	})
+
+	It("does not overflow the time.Duration type", func() {
+		strategy := Linear(linger.MaxDuration / 2)
+
+		Expect(strategy(nil, 0)).To(Equal(linger.MaxDuration / 2))
+		Expect(strategy(nil, 1)).To(Equal(linger.MaxDuration - 1))
+		Expect(strategy(nil, 2)).To(Equal(linger.MaxDuration))
+		Expect(strategy(nil, 3)).To(Equal(linger.MaxDuration))
+		Expect(strategy(nil, 4)).To(Equal(linger.MaxDuration))
 	})
 })
 

--- a/backoff/strategy_test.go
+++ b/backoff/strategy_test.go
@@ -1,6 +1,7 @@
 package backoff_test
 
 import (
+	"math"
 	"time"
 
 	"github.com/dogmatiq/linger"
@@ -32,8 +33,14 @@ var _ = Describe("func Exponential()", func() {
 	It("does not overflow the time.Duration type", func() {
 		strategy := Exponential(1)
 
-		Expect(strategy(nil, 62)).To(Equal(time.Duration(4611686018427387904)))
+		// No overflow at 2^62. We have 63-bits of positive magnitude available
+		// because time.Duration is a signed 64-bit integer.
+		Expect(strategy(nil, 62)).To(Equal(time.Duration(math.Pow(2, 62))))
+
+		// Starts overflowing at 2^63.
 		Expect(strategy(nil, 63)).To(Equal(linger.MaxDuration))
+
+		// Continues to return the capped value as the exponent increases.
 		Expect(strategy(nil, 100)).To(Equal(linger.MaxDuration))
 	})
 
@@ -69,12 +76,21 @@ var _ = Describe("func Linear()", func() {
 	})
 
 	It("does not overflow the time.Duration type", func() {
-		strategy := Linear(linger.MaxDuration / 2)
+		unit := linger.MaxDuration - 1
 
-		Expect(strategy(nil, 0)).To(Equal(linger.MaxDuration / 2))
-		Expect(strategy(nil, 1)).To(Equal(linger.MaxDuration - 1))
-		Expect(strategy(nil, 2)).To(Equal(linger.MaxDuration))
-		Expect(strategy(nil, 3)).To(Equal(linger.MaxDuration))
+		strategy := Linear(unit)
+
+		// No overflow at 1 * unit, as it's slightly below the max.
+		Expect(strategy(nil, 0)).To(Equal(unit))
+
+		// Start overflowing at 2 * unit. Overflow is detected because the
+		// result of the multiplication wraps negative.
+		Expect(unit * 2).To(BeNumerically("<", 0)) // verify test inputs actually wrap
+		Expect(strategy(nil, 1)).To(Equal(linger.MaxDuration))
+
+		// Continue overflowing. Overflow is detected because the result of the
+		// multiplication wraps negative, and continues on to be positive again.
+		Expect(unit * 5).To(BeNumerically(">", 0)) // verify test inputs actually wrap
 		Expect(strategy(nil, 4)).To(Equal(linger.MaxDuration))
 	})
 })

